pkg/controllers/ipam/ipam.go

```diff
@@ -429,6 +430,18 @@ func (i *Manager) AssignIP(ctx context.Context, allocate *svcRecord) error {
 		return err
 	}
 
+	parseRange, err := iprange.ParseRange(eip.Spec.Address)
+	if err != nil {
+		return err
+	}
+	eipFamily := parseRange.Family()
+	if svc.Spec.IPFamilies[0] == "IPv4" && eipFamily == iprange.V6Family {
+		return fmt.Errorf("ipv4 service can't use ipv6 eip")
+	}
+	if svc.Spec.IPFamilies[0] == "IPv6" && eipFamily == iprange.V4Family {
+		return fmt.Errorf("ipv6 service can't use ipv4 eip")
+	}
```

Comment:
You only considered the single stack case, not the dual stack case.

When the service is dual stack, svc.Spec.IPFamilies contains multiple elements ["IPV4","IPV6"]